plugin/pkg/replacer/replacer.go

```diff
@@ -36,80 +41,92 @@ type replacer struct {
+			"{port}":   func() string { return req.Port() },
+			"{local}":  func() string { return addrToRFC3986(req.LocalIP()) },
 		},
+		evaluated:  make(map[string]string, 0),
 		emptyValue: emptyValue,
 	}
 	if rr != nil {
 		rcode := dns.RcodeToString[rr.Rcode]
 		if rcode == "" {
 			rcode = strconv.Itoa(rr.Rcode)
 		}
-		rep.replacements["{rcode}"] = rcode
-		rep.replacements["{rsize}"] = strconv.Itoa(rr.Len)
-		rep.replacements["{duration}"] = strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s"
+		rep.replacements["{rcode}"] = func() string { return rcode }
+		rep.replacements["{rsize}"] = func() string { return strconv.Itoa(rr.Len) }
+		rep.replacements["{duration}"] = func() string { return strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s" }
 		if rr.Msg != nil {
-			rep.replacements[headerReplacer+"rflags}"] = flagsToString(rr.Msg.MsgHdr)
-			rep.replacements["{A}"] = answersCount(rr.Msg, dns.TypeA)
-			rep.replacements["{AAAA}"] = answersCount(rr.Msg, dns.TypeAAAA)
+			rep.replacements[headerReplacer+"rflags}"] = func() string { return flagsToString(rr.Msg.MsgHdr) }
 		}
 	}
 
 	// Header placeholders (case-insensitive)
-	rep.replacements[headerReplacer+"id}"] = strconv.Itoa(int(r.Id))
-	rep.replacements[headerReplacer+"opcode}"] = strconv.Itoa(r.Opcode)
-	rep.replacements[headerReplacer+"do}"] = boolToString(req.Do())
-	rep.replacements[headerReplacer+"bufsize}"] = strconv.Itoa(req.Size())
+	rep.replacements[headerReplacer+"id}"] = func() string { return strconv.Itoa(int(r.Id)) }
+	rep.replacements[headerReplacer+"opcode}"] = func() string { return strconv.Itoa(r.Opcode) }
+	rep.replacements[headerReplacer+"do}"] = func() string { return boolToString(req.Do()) }
+	rep.replacements[headerReplacer+"bufsize}"] = func() string { return strconv.Itoa(req.Size()) }
 
 	return rep
 }
 
 // Replace performs a replacement of values on s and returns
 // the string with the replaced values.
 func (r replacer) Replace(s string) string {
-	// Header replacements - these are case-insensitive, so we can't just use strings.Replace()
-	for strings.Contains(s, headerReplacer) {
-		idxStart := strings.Index(s, headerReplacer)
-		endOffset := idxStart + len(headerReplacer)
+	b := strings.Builder{}
+	for strings.Contains(s, "{") {
+		idxStart := strings.Index(s, "{")
+		endOffset := idxStart + 1
 		idxEnd := strings.Index(s[endOffset:], "}")
 		if idxEnd > -1 {
 			placeholder := strings.ToLower(s[idxStart : endOffset+idxEnd+1])
-			replacement := r.replacements[placeholder]
+			replacement, ok := r.evaluated[placeholder]
+			if !ok {
+				f, fok := r.replacements[placeholder]
+				if fok {
+					replacement = f()
+				} else {
+					label := placeholder[1 : len(placeholder)-1]
+					fm := metadata.ValueFunc(r.ctx, label)
+					if fm != nil {
+						replacement = fm()
+					} else {
+						if !metadata.IsLabel(label) {
+							replacement = "<unknown-place-holder>"
```

Comment:
this is not `emptyValue`

Reply:
if the placeholder is not a standard and is has no format of a metadata label, then it is obviously an error in the Corefile. Here is a way to warn it.

Reply:
well, you get EmptyValue then or something. If you want to warn a person you probably need to crash the process; no-one will see this or look for it in a running environment.